File: C++/tests/binary_tree_test.cc
#include "../lib/binary_tree.hpp"
#include <gtest/gtest.h>

TEST(BinaryTreeTest, from_method_case1) {
  TreeNode *root = TreeNode::from({1, 2, 3});
  TreeNode *expected = new TreeNode(1, new TreeNode(2), new TreeNode(3));

  EXPECT_EQ(*root, *expected);

  free_tree(root);
  free_tree(expected);
}

TEST(BinaryTreeTest, from_method_case2) {
  TreeNode *root = TreeNode::from({1, 2, 3, 4});
  TreeNode *expected = new TreeNode(
      1, new TreeNode(2, new TreeNode(4), nullptr), new TreeNode(3));

  EXPECT_EQ(*root, *expected);

  free_tree(root);
  free_tree(expected);
}

TEST(BinaryTreeTest, from_method_case3) {
  TreeNode *root = TreeNode::from({1, 2, 3, 4, 5, 6, 7});
  TreeNode *expected =
      new TreeNode(1, new TreeNode(2, new TreeNode(4), new TreeNode(5)),
                   new TreeNode(3, new TreeNode(6), new TreeNode(7)));

  EXPECT_EQ(*root, *expected);

  free_tree(root);
  free_tree(expected);
}

TEST(BinaryTreeTest, from_method_case4) {
  TreeNode *root = TreeNode::from({1, 2, nullptr, 3, nullptr, 4});
  TreeNode *expected = new TreeNode(
      1, new TreeNode(2, new TreeNode(3, new TreeNode(4), nullptr), nullptr),
      nullptr);

  EXPECT_EQ(*root, *expected);

  free_tree(root);
  free_tree(expected);
}

TEST(BinaryTreeTest, from_method_case5) {
  TreeNode *root = TreeNode::from({1, nullptr, 2, nullptr, 3, nullptr, 4});
  TreeNode *expected = new TreeNode(
      1, nullptr,
      new TreeNode(2, nullptr, new TreeNode(3, nullptr, new TreeNode(4))));

  EXPECT_EQ(*root, *expected);

  free_tree(root);
  free_tree(expected);
}

TEST(BinaryTreeTest, from_method_case6) {
  TreeNode *root = TreeNode::from({1, 2, 3, nullptr, 5, 6});
  TreeNode *expected =
      new TreeNode(1, new TreeNode(2, nullptr, new TreeNode(5)),
                   new TreeNode(3, new TreeNode(6), nullptr));

  EXPECT_EQ(*root, *expected);

  free_tree(root);
  free_tree(expected);
}

TEST(BinaryTreeTest, from_method_case7) {
  TreeNode *root = TreeNode::from({});
  EXPECT_EQ(root, nullptr);
}

TEST(BinaryTreeTest, from_method_case8) {
  TreeNode *root = TreeNode::from({1});
  TreeNode *expected = new TreeNode(1);

  EXPECT_EQ(*root, *expected);

  free_tree(root);
  free_tree(expected);
}

File: JS/tests/binary_tree.test.js
import assert from "node:assert";
import { describe, it } from "node:test";
import { TreeNode } from "../lib/binary_tree.js";

describe("TreeNode", { concurrency: true, timeout: 1000 }, () => {
  it("should create correct Tree with [0, null, 1, 2]", () => {
    const tree = TreeNode.of(0, null, 1, 2);
    const expected = new TreeNode(0, null, new TreeNode(1, new TreeNode(2)));

    assert.deepStrictEqual(tree, expected);
  });

  it("should create correct Tree with [1, 2, 3, null, null, 4, 5]", () => {
    const tree = TreeNode.of(1, 2, 3, null, null, 4, 5);
    const expected = new TreeNode(
      1,
      new TreeNode(2),
      new TreeNode(3, new TreeNode(4), new TreeNode(5)),
    );

    assert.deepStrictEqual(tree, expected);
  });

  it("should create correct Tree with [1, null, 2, 3]", () => {
    const tree = TreeNode.of(1, null, 2, 3);
    const expected = new TreeNode(1, null, new TreeNode(2, new TreeNode(3)));

    assert.deepStrictEqual(tree, expected);
  });

  it("should create correct Tree with [1, 2, null, 3, null, 4, null]", () => {
    const tree = TreeNode.of(1, 2, null, 3, null, 4, null);
    const expected = new TreeNode(
      1,
      new TreeNode(2, new TreeNode(3, new TreeNode(4))),
    );

    assert.deepStrictEqual(tree, expected);
  });

  it("should create correct Tree with [1]", () => {
    const tree = TreeNode.of(1);
    const expected = new TreeNode(1);

    assert.deepStrictEqual(tree, expected);
  });

  it("should return null for an empty array", () => {
    const tree = TreeNode.of();
    const expected = null;

    assert.deepStrictEqual(tree, expected);
  });
});
